File: internal/fingerprint/fingerprint.go
// Package fingerprint decides what "changed" means for each kind of object: a
// module hashes its definition, a table its structure plus its sample depth.
// Sample row VALUES are deliberately absent — they move on every run against a
// live database and would leave every table permanently dirty.
package fingerprint

import (
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"strings"
	"unicode"

	"github.com/branow/dbmap/internal/catalog"
)

// SampleRows is how many rows the describer is sent; 25 comes from the measured
// row-count distribution (docs/DESIGN.md).
const SampleRows = 25

// digits keeps 64 bits of the hash, short enough to sit in a TSV row.
const digits = 16

// SampleDepth is how many rows the describer gets to see, so a table growing
// past SampleRows does not redescribe.
func SampleDepth(rows int64) int64 {
	if rows < 0 {
		return 0
	}
	if rows > SampleRows {
		return SampleRows
	}
	return rows
}

// Sum is the hash every fingerprint here is taken with.
func Sum(text string) string {
	sum := sha256.Sum256([]byte(text))
	return hex.EncodeToString(sum[:])[:digits]
}

// Normalize strips line endings and trailing whitespace, so a deployment tool
// rewriting a file to CRLF does not read as an edit.
func Normalize(definition string) string {
	lines := strings.Split(strings.ReplaceAll(definition, "\r\n", "\n"), "\n")
	for i, line := range lines {
		lines[i] = strings.TrimRightFunc(line, unicode.IsSpace)
	}
	return strings.TrimSpace(strings.Join(lines, "\n"))
}

// Module fingerprints a view, procedure or function over its body.
func Module(definition string) string { return Sum(Normalize(definition)) }

// CanonicalTable is the exact text a table's fingerprint is taken over. Written
// out line by line so adding a field to Structure cannot move the hash.
func CanonicalTable(object catalog.Object, structure catalog.Structure) string {
	lines := make([]string, 0, len(structure.Columns)+len(structure.Indexes)+4)
	for _, c := range structure.Columns {
		line := "c " + c.Name + " " + c.Type + c.Length + " " + nullability(c.Nullable)
		if c.Identity {
			line += " id"
		}
		lines = append(lines, line)
	}
	for _, k := range structure.PrimaryKey {
		lines = append(lines, "pk "+k)
	}
	for _, f := range structure.ForeignKeys {
		lines = append(lines, "fk "+f.Column+" -> "+f.References)
	}
	for _, i := range structure.Indexes {
		line := "ix " + i.Name + " " + strings.Join(i.Columns, ",")
		if i.Unique {
			line += " u"
		}
		lines = append(lines, line)
	}
	lines = append(lines, "tg "+strconv.Itoa(object.Triggers))
	lines = append(lines, "rows "+strconv.FormatInt(SampleDepth(object.Rows), 10))
	return strings.Join(lines, "\n")
}

func nullability(nullable bool) string {
	if nullable {
		return "null"
	}
	return "notnull"
}

// Table fingerprints a table over its canonical structure.
func Table(object catalog.Object, structure catalog.Structure) string {
	return Sum(CanonicalTable(object, structure))
}

func module(_ catalog.Object, structure catalog.Structure) string {
	return Module(structure.Definition)
}

func synonym(_ catalog.Object, structure catalog.Structure) string {
	return Sum(structure.Target)
}

var fingerprinters = map[catalog.Kind]func(catalog.Object, catalog.Structure) string{
	catalog.Table:     Table,
	catalog.View:      module,
	catalog.Procedure: module,
	catalog.Function:  module,
	catalog.Synonym:   synonym,
}

// Of returns the content fingerprint for one object. A kind with no rule is
// refused: a guessed hash either redescribes forever or never again.
func Of(object catalog.Object, structure catalog.Structure) (string, error) {
	fn, ok := fingerprinters[object.Kind]
	if !ok {
		return "", UnknownKindError{Kind: object.Kind}
	}
	return fn(object, structure), nil
}

// UnknownKindError reports a kind the fingerprint table has no rule for.
type UnknownKindError struct {
	Kind catalog.Kind
}

func (e UnknownKindError) Error() string {
	return "fingerprint: no rule for kind " + string(e.Kind)
}

File: reference/js/fingerprint.js
"use strict";
// What "changed" means for each kind of object.
//
// Two different signals guard two different costs:
//
//   modify_date  decides whether to FETCH. It is free (already in the manifest)
//                and never misses a real change, but it over-reports: a release
//                that ALTERs 58 procedures bumps all 58 dates even when most are
//                byte-identical to what was already deployed.
//   fingerprint  decides whether to DESCRIBE. Fetching is seconds; describing is
//                an LLM call per object, so the exact check guards the expensive
//                stage and the loose one guards the cheap stage.
//
// A module fingerprints over its definition. A table has no definition, so it
// fingerprints over the structure a reader would need — columns with their
// types and nullability, keys, indexes, trigger count — plus its sample depth.
//
// Sample row VALUES are deliberately absent. They change on every run against a
// live database and would leave every table permanently dirty.

const { createHash } = require("node:crypto");

// Rows sent to the describer, and therefore also the row count above which a
// table growing tells the describer nothing new: TOP 25 of a 40-row table and
// TOP 25 of a 60-row table are the same 25 rows.
//
// 25 because 148 of the 491 non-empty tables on stage hold 25 rows or fewer —
// the lookup and enum tables, where every row is a value worth naming and the
// sample is the whole table. Past 25 the distribution flattens: raising it to 30
// covers five more tables.
const SAMPLE_ROWS = 25;

// How many rows the describer actually gets to see. Below SAMPLE_ROWS a table
// growing by one row changes what the model reads, so the description is stale;
// at or above it, growth changes nothing.
function sampleDepth(rows) {
  return Math.min(Number(rows) || 0, SAMPLE_ROWS);
}

function sha256(text) {
  return createHash("sha256").update(text, "utf8").digest("hex").slice(0, 16);
}

// Line endings and trailing whitespace differ between deployment tools without
// the code differing, so they are normalised away before hashing.
function normalizeDefinition(definition) {
  return String(definition ?? "")
    .replace(/\r\n/g, "\n")
    .split("\n")
    .map((line) => line.replace(/\s+$/, ""))
    .join("\n")
    .trim();
}

function moduleFingerprint(definition) {
  return sha256(normalizeDefinition(definition));
}

// Canonical text for a table's structure. Built as explicit lines rather than
// JSON so the hash does not move when an unrelated field is added to the
// structure objects the fetch stage happens to return.
function canonicalTable(structure) {
  const lines = [];
  for (const c of structure.columns ?? [])
    lines.push(`c ${c.name} ${c.type}${c.length ?? ""} ${c.nullable ? "null" : "notnull"}${c.identity ? " id" : ""}`);
  for (const k of structure.primaryKey ?? []) lines.push(`pk ${k}`);
  for (const f of structure.foreignKeys ?? []) lines.push(`fk ${f.column} -> ${f.references}`);
  for (const i of structure.indexes ?? [])
    lines.push(`ix ${i.name} ${(i.columns ?? []).join(",")}${i.unique ? " u" : ""}`);
  lines.push(`tg ${structure.triggers ?? 0}`);
  lines.push(`rows ${sampleDepth(structure.rows)}`);
  return lines.join("\n");
}

function tableFingerprint(structure) {
  return sha256(canonicalTable(structure));
}

// Dispatches on kind so callers never branch on it themselves.
const FINGERPRINTERS = {
  table: tableFingerprint,
  view: (s) => moduleFingerprint(s.definition),
  procedure: (s) => moduleFingerprint(s.definition),
  function: (s) => moduleFingerprint(s.definition),
  synonym: (s) => sha256(String(s.target ?? "")),
};

function fingerprint(kind, payload) {
  const fn = FINGERPRINTERS[kind];
  if (!fn) throw new Error(`db-index: no fingerprint defined for kind "${kind}"`);
  return fn(payload ?? {});
}

module.exports = {
  SAMPLE_ROWS,
  sampleDepth,
  normalizeDefinition,
  moduleFingerprint,
  canonicalTable,
  tableFingerprint,
  fingerprint,
  sha256,
};
